File: code/java/unit-work/src/main/java/org/swufe/datastructure/BinarySearch.java
package org.swufe.datastructure;
import java.util.List;

public class BinarySearch {
    /**
     * Binary Search
     * @param list a sorted List
     * @return the index of `target` (-1 if not found)
     */
    public int search(List<Integer> list, int target) {
        int high = list.size() - 1;
        int low = 0;
        while (high >= low) {
            int mid = low + (high - low) / 2;
            if (list.get(mid) == target) {
                return mid;
            } else if (list.get(mid) > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }
}

File: code/python/start/binary_search.py
def search(a, target):
    """
    Binary search.

    :param a: a sorted list
    :param target
    :return: the index of `target` (-1 if not found)
    """
    high = len(a) - 1
    low = 0
    while high >= low:
        mid = low + (high - low) // 2
        if a[mid] == target:
            return mid
        elif a[mid] > target:
            high = mid - 1
        else:
            low = mid + 1
    return -1
